lib/internal/streams/state.js

```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+const stream = require('stream');
+const errors = require('internal/errors');
+
+function getHighWaterMarkFromProperty(hwm, name) {
+  if (typeof hwm !== 'number' || hwm < 0)
+    throw new errors.TypeError('ERR_INVALID_OPT_VALUE', name, hwm);
+  return Math.floor(hwm);
+}
+
+function getHighWaterMark(state, options, isDuplex) {
+  if (options.highWaterMark != null && !Number.isNaN(options.highWaterMark)) {
+    return getHighWaterMarkFromProperty(options.highWaterMark, 'highWaterMark');
+  } else if (isDuplex) {
+    if (state instanceof stream.Readable.ReadableState) {
+      if (options.readableHighWaterMark != null &&
+          !Number.isNaN(options.readableHighWaterMark)) {
+        return getHighWaterMarkFromProperty(options.readableHighWaterMark,
+                                            'readableHighWaterMark');
+      }
+    } else if (options.writableHighWaterMark != null &&
+               !Number.isNaN(options.writableHighWaterMark)) {
+      return getHighWaterMarkFromProperty(options.writableHighWaterMark,
+                                          'writableHighWaterMark');
+    }
+  }
+
+  // Default value
+  return state.objectMode ? 16 : 16 * 1024;
+}
```

Comment:
The following should be way more efficient (untested):
```js
function getHighWaterMark (options, isDuplex, specialKey, objectMode) {
  if (typeof options.highWaterMark === 'number') {
    if (options.highWaterMark < 0 || Number.isNaN(options.highWaterMark)) {
      throw new errors.TypeError(
        'ERR_INVALID_OPT_VALUE',
        'options.highWaterMark',
        options.highWaterMark
      )
    }
    return Math.floor(options.highWaterMark)
  }
  if (isDuplex) {
    if (typeof options[specialKey] === 'number') {
      if (options[specialKey] < 0 || Number.isNaN(options[specialKey])) {
        throw new TypeError(...)
      }
      return Math.floor(options[specialKey])
    }
    if (options[specialKey]) {
      throw new TypeError(...)
    }
  } else if(options.highWaterMark) {
    throw new TypeError(...)
  }

  return objectMode ? 16 : 16 * 1024
}

getHighWaterMark(
  options,
  isDuplex,
  'writableHighWaterMark', // Replace accordingly
  this.objectMode
)
```

Reply:
Yeah I was going to extract the key anyway, just wanted to hear opinions about `NaN` first (that really annoyed me).

Reply:
By the way, `!(options.highWaterMark >= 0)` should be faster than `options.highWaterMark < 0 || Number.isNaN(options.highWaterMark)`.

Reply:
That is probably true.